### PR TITLE
add emit log in auditlog.js, add ringbuffer to store certain amount of logs

### PR DESCRIPTION
**Pull request by leitsubomi**
Add emit log in auditlog.js
Add ringbuffer to store certain amount of logs logMetrics.js
Unit tests plugin.test.js
Add unit test in server.test.js to verify restify behavior when user creates duplicate route path.

### CODE REVIEW

File: lib/plugins/audit.js [hunk 1]
```diff
@@ -25,6 +25,10 @@ var HttpError = require('restify-errors').HttpError;
 function auditLogger(options) {
     assert.object(options, 'options');
     assert.object(options.log, 'options.log');
```
Comment (yunong) [inline lib/plugins/audit.js]: Assert for the server object here.
```js
assert.object(options.server, 'options.server');

File: lib/plugins/audit.js [hunk 2]
```diff
@@ -25,6 +25,10 @@ var HttpError = require('restify-errors').HttpError;
 function auditLogger(options) {
```
Comment (yunong) [inline lib/plugins/audit.js]: Update the docs to reflect your API changes.

File: lib/plugins/audit.js [hunk 3]
```diff
@@ -110,6 +115,14 @@ function auditLogger(options) {
 
         log.info(obj, 'handled: %d', res.statusCode);
 
+        if (logMetrics && logMetrics instanceof Object) {
```
Comment (yunong) [inline lib/plugins/audit.js]: You don't need these checks here if you just assert like the other objects up on line 27

File: lib/plugins/audit.js [hunk 4]
```diff
@@ -110,6 +115,14 @@ function auditLogger(options) {
...
+            obj.timestamp = timestamp;
+            logMetrics.pushLog(obj);
+        }
+
+        if (server && server instanceof Object) {
```
Comment (yunong) [inline lib/plugins/audit.js]: same here, no need to check, just assert this object up top.

File: lib/plugins/logMetrics.js [hunk 5]
```diff
@@ -0,0 +1,33 @@
+'use strict';
+
+var bunyan = require('bunyan');
+
+function logMetrics(options) {
```
Comment (yunong) [inline lib/plugins/logMetrics.js]: Classes should start with a capital letter. Let's call this MetricBuffer instead of logMetrics.

Comment (yunong) [inline lib/plugins/logMetrics.js]: Also add jsdocs about this class.

File: lib/plugins/audit.js [hunk 6]
```diff
@@ -25,6 +25,10 @@ var HttpError = require('restify-errors').HttpError;
 function auditLogger(options) {
     assert.object(options, 'options');
     assert.object(options.log, 'options.log');
+    //use server object to emit log data
+    var server = options.server;
+    //use logMetrics object to store certain period of log records
+    var logMetrics = options.logMetrics;
```
Comment (yunong) [inline lib/plugins/audit.js]: The LogMetric class is overkill. Just have the user pass in a ringbuffer object themselves, and directly use that instead.

File: lib/plugins/audit.js (hunk 2 above)
Comment (yunong) [inline lib/plugins/audit.js]: Since this class is no longer just writing logs, change its name to `audit`  instead.

File: test/plugins.test.js [hunk 7]
```diff
@@ -6,9 +6,12 @@ var fs = require('fs');
 var http = require('http');
 var net = require('net');
 var path = require('path');
+var util = require('util');
 
 var bunyan = require('bunyan');
 var restifyClients = require('restify-clients');
+var assert = require('chai').assert;
```
Comment (yunong) [inline test/plugins.test.js]: Please sort imports by alphabetical order.

File: lib/plugins/audit.js [hunk 8]
```diff
@@ -91,6 +95,7 @@ function auditLogger(options) {
 
     function audit(req, res, route, err) {
         var latency = res.get('Response-Time');
+        var timestamp = new Date().toString();
```
Comment (yunong) [inline lib/plugins/audit.js]: Don't use the fully qualified date object here, this will make it hard to parse. Instead, give it ns since epoch

File: lib/plugins/audit.js [hunk 9]
```diff
@@ -20,11 +20,29 @@ var HttpError = require('restify-errors').HttpError;
...
+    assert.optionalObject(options.logBuffer, 'options.logBuffer');
+    assert.optionalBool(options.printLog, 'options.printLog');
+    //use server object to emit log data
+    var server = options.server;
+    //use logMetrics ringbuffer to store certain period of log records
+    var logMetrics = options.logBuffer;
+    //default always print log
+    var printLog = options.printLog;
+
+    if (typeof printLog === 'undefined') {
```
Comment (yunong) [inline lib/plugins/audit.js]: this can just be `if (!printlog)`

**Comment by yunong** [comment]
@leitsubomi Looks great. Can you update CHANGES.md with your changes as well as docs/index.restdown?

**Comment by DonutEspresso** [comment]
Looking at this PR, I think we actually want this merged in against restify/plugins. That means we don't have to tie it in with the 5.x release. 

The tests for duplicate routes can probably be refiled as a separate PR, still against this repo.
